src/core/utils/AlbedoNormalPass.js

```diff
@@ -0,0 +1,224 @@
+                uniform sampler2D normalMap;
+                uniform bool useWorldSpaceNormals;
+                uniform bool useDiffuseMap;
+                uniform bool useNormalMap;
+                uniform bool debugNormalMaps;
+				uniform float normalScale;
+
+                varying vec2 vUv;
+                varying vec3 vViewPosition;
+                varying vec3 vNormal;
+                varying vec3 vWorldNormal;
+                varying vec3 vTangent;
+                varying vec3 vBitangent;
+
+                layout(location = 0) out vec4 gAlbedo;
+                layout(location = 1) out vec4 gNormal;
+
+                void main() {
+                    // Albedo
+                    vec3 albedo;
+                    if (useDiffuseMap) {
+                        vec4 diffuseColor = texture(diffuseMap, vUv);
+                        albedo = diffuseColor.rgb * color;
+                    } else {
+                        albedo = color;
+                    }
+                    gAlbedo = vec4(albedo, 1.0);
+
+                    // Normal
+                    vec3 normal;
+                    if (useNormalMap) {
+						// This is NOT correct, but it works. I dont think the denoiser actually cares
+						// TODO: Make this actually work to convert to worldSpace
+						// Output the raw normal map data
+						vec3 rawNormalMap = texture(normalMap, vUv).xyz;
+						gNormal = vec4(rawNormalMap, 1.0);
+						return;
+					
+						vec3 mapN = texture(normalMap, vUv).xyz * 2.0 - 1.0;
+						mapN.xy *= normalScale;
+						mat3 tbn = mat3(normalize(vTangent), normalize(vBitangent), normalize(vNormal));
+						vec3 worldNormal = normalize(tbn * mapN);
+						
+						if (useWorldSpaceNormals) {
+							normal = worldNormal;
+						} else {
+							normal = normalize((viewMatrix * vec4(worldNormal, 0.0)).xyz);
+						}
+					} else {
+						normal = useWorldSpaceNormals ? vWorldNormal : vNormal;
+					}
+
+					normal = normalize(normal + vec3(1e-6));
+					normal = normal * 0.5 + 0.5;
+					gNormal = vec4(normal, 1.0);
+                }
+            `,
+			glslVersion: THREE.GLSL3
+		} );
+
+	}
+
+	async render( renderer, scene, camera ) {
+
+		const target = this.renderTarget;
+		target.setSize( renderer.domElement.width, renderer.domElement.height );
+
+		this.swapMaterials( scene );
+
+		const oldRenderTarget = renderer.getRenderTarget();
+		renderer.setRenderTarget( target );
+		renderer.render( scene, camera );
+		renderer.setRenderTarget( oldRenderTarget );
+
+		this.restoreMaterials();
+
+		// return the two textures
+		return { albedo: target.textures[ 0 ], normal: target.textures[ 1 ] };
+
+	}
+
+	swapMaterials( object ) {
+
+		if ( object instanceof THREE.Mesh && object.material ) {
+
+			if ( ! this.originalMaterials.has( object ) )
+				this.originalMaterials.set( object, object.material );
+
+			if ( this.albedoMaterials.has( object ) ) {
+
+				object.material = this.albedoMaterials.get( object );
+				return;
+
+			}
```

Comment:
I just mean you keep a pool of materials and assign all the necessary material properties for each object when assigning:

```js
// before
meshes.forEach( mesh => {

  const material = materialPool.getMaterial();
  material.normalMap = mesh.material.normalMap;
  material.map = mesh.material.map;
  
  originalMaterials.set( mesh, mesh.material );
  mesh.material = material;

} );

// after
meshes.forEach( mesh => {

  mesh.material.normalMap = null;
  mesh.material.map = null;
  mesh.material = originalMaterials.get( mesh );

} );

originalMaterials.clear();
materialPool.reset();
```

This way you avoid keeping textures around on hidden internal material objects when the user has intended to get rid of them. And likewise you avoid using the Mesh object as a key which will also keep a reference around preventing it from being GC'd when the user has removed all references.

Reply:
In a different message you mentioned doing it in two passes being ok. And as we aren't doing it every frame I went ahead and went that route. Also the normals dont actually have to be physically correct, it's just to help the denoiser demark edges.

So I rewrote this whole thing.

Now we dont maintain anything. All the materials are stored on the mesh itself. So if it gets destroyed they go with it.
I share the plain normal material and clone when a normal map is present. 

The albedo shader is my first pass at responding to different material settings to set albedo instead of just a pure color. Its being weird for some parts so this is where I'll focus next.

I'll mention this more in my post

Reply:
The code looks a lot cleaner by using the built in materials - nice work!

> So if it gets destroyed they go with it.

There is still a memory leak here since the materials on `userData` don't get disposed when an object GC'd so there will be dangling shader references left in the three.js renderer.

I think I still prefer the pool method since it's more self-contained and the pooled materials can be properly disposed when the path tracer and / or denoiser is disposed.

Reply:
Gotcha, I thought putting it in userdata would track the reference. I'll work on this now